### PR TITLE
Creating_parse_bar_color_args to unify color handling in plt.bar with precedence and sequence support for facecolor and edgecolor

### LINKED ISSUE #29090: [MNT]: More consistent color parameters for bar()
### Summary

From #29072. `bar()` supports

- `color` : color or list of color
- `edgecolor` : color or list of color
- `facecolor`: color

i.e.

- `facecolor` cannot take a sequence
- there are no plural aliase (e.g. `edgecolors`)
- likely (t.b.c.) the aliases also do not support sequences, similar to #28884

### Proposed fix

Make `facecolor` accept sequences and check that the parameter precedence among `color`, `edgecolor` and `facecolor` is reasonable and comparable with `scatter`, which can take an explicit color via `c` (equivalent to `color` here).

For now, I'd refrain from introducing plural aliase. `bar()` is originally and primarily a style-all-bars-identical function. Per-bar styling was added later on, and I don't think there's a strong need to support this additional use case with an added plural alias.

### PR DESCRIPTION
**Pull request by Lucx33**
## PR summary

This PR introduces the `_parse_bar_color_args function`, which provides consistency in color bar arguments by establishing a parameter precedence, "similar" to the one used in `scatter`:

facecolor: facecolor -> color -> 'b' if in classic mode else the result of `get_next_color_func()`

edgecolor: edgecolor -> color -> None

This ensures that when edgecolors is not explicitly defined, it tries color. Additionally, this change allows facecolor to accept sequences of colors, ensuring consistency in argument handling, as issued in #29090. The tests are consistent with examples in #29072.

**Examples:**

- `plt.bar([1, 2, 3], [1, 2, 3], color=['none', 'C2', 'none'], edgecolor=['C1', 'C2', 'C3'])`
Figure_1
 
- `plt.bar([1, 2, 3], [1, 2, 3], color=['none', 'C2', 'none'])`
Figure_2

- `plt.bar([1, 2, 3], [1, 2, 3])
`Figure_3

- `plt.bar([1, 2, 3], [1, 2, 3],facecolor=('none', 'C2', 'none'))`
Figure_4

- `plt.bar([1, 2, 3], [1, 2, 3], color=('C1', 'C2', 'C3'), facecolor=('none', 'C2', 'none'))`
Figure_5

## PR checklist

- [x] "closes #29090" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by Lucx33** [comment]
Although some tests are failing, I can't determine the cause; For example, test_bbox_inches_tight in test_bbox_tight fails, even though old/new images appear equal.

**Comment by rcomer** [comment]
The easiest test to debug might be `test_bar_color_cycle` in `test_axes.py` since that one explicitly shows that we are expecting blue but getting green.  So, having already plotted a line, `bar` is not using the first color in the cycle.

File: lib/matplotlib/axes/_axes.py
Comment (rcomer) [inline lib/matplotlib/axes/_axes.py:2364]: I guess you made "b" the default in classic mode for consistency with `scatter`.  However this is not the established behaviour for `bar` and setting it to blue is likely the cause of at least some of your test failures.  For `scatter` it will only be there for consistency with old Matplotlib versions.  I think we should just use the `get_next_color_func` here.

File: lib/matplotlib/axes/_axes.py
Comment (rcomer) [inline lib/matplotlib/axes/_axes.py:2325]: According to #12739 the method for parsing `scatter` color arguments was made static so that we could have unit tests specifically for that method.  Do you plan to write similar tests here?  If not, this could be an ordinary method and we would not need to pass the `get_next_color_func` but just access `self._get_patches_for_fill.get_next_color` directly.

**Comment by Lucx33** [comment]
@rcomer thx for the review, I'll fix it as soon as possible. prob later today

**Comment by Lucx33** [comment]
@rcomer  test_bar_color_cycle passed as you said. Others still failed.

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2353]:
```suggestion
            The facecolor. One or more colors as (N, 4) rgba array.
```

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2355]:
```suggestion
            The edgecolor. Not normalized; may be any valid color spec or None.
```

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2498]:
```suggestion
        facecolor, edgecolor = self._parse_bar_color_args(kwargs)
```
Let's make this explicit and change the naming from here on - needs updates below.

Also note: It's used below in
```
color = itertools.chain(itertools.cycle(mcolors.to_rgba_array(color))
```
You can now leave out the `to_rgba_array()` there, because your helper already does the conversion.

**Comment by rcomer** [comment]
Here are the before or after of that `test_bbox_inches_tight` according to the tests.  Note that the image tests run in classic mode unless [the style keyword is explicitly set](https://matplotlib.org/devdocs/devel/testing.html#writing-an-image-comparison-test).

Before
bbox_inches_tight-expected


After
bbox_inches_tight

Difference
bbox_inches_tight-failed-diff


So it seems the edgecolor has changed from black to blue.

**Note** you can download all the test images under "Artifacts" on the [Test Summary Page](https://github.com/matplotlib/matplotlib/actions/runs/11846448112?pr=29133).  However those zip files are large so it may be more efficient to run the tests locally and inspect the images you produce that way.

**Comment by Lucx33** [comment]
@rcomer  In `test_bbox_inches_tight`, since our new precedence says that if color is specified, it defines both facecolor and edgecolor ('b' in this case). Changing color to facecolor fixes the issue:
Old: `ax.bar(ind, data[row], width, bottom=yoff, align='edge', color='b')`
New: `ax.bar(ind, data[row], width, bottom=yoff, align='edge', facecolor='b')`

> Here are the before and after of that `test_bbox_inches_tight` according to the tests. Note that the image tests run in classic mode unless [the style keyword is explicitly set](https://matplotlib.org/devdocs/devel/testing.html#writing-an-image-comparison-test).
> 
> Before bbox_inches_tight-expected
> 
> After bbox_inches_tight
> 
> Difference bbox_inches_tight-failed-diff
> 
> So it seems the edgecolor has changed from black to blue.
> 
> **Note** you can download all the test images under "Artifacts" on the [Test Summary Page](https://github.com/matplotlib/matplotlib/actions/runs/11846448112?pr=29133). However those zip files are large so it may be more efficient to run the tests locally and inspect the images you produce that way.

**Comment by Lucx33** [comment]
Other tests fails for the same reason as both tests above.

**Comment by rcomer** [comment]
I do not think we really want _edgecolor_ to match _color_ by default:  if I take [the grouped bar example](https://matplotlib.org/stable/gallery/lines_bars_and_markers/barchart.html#sphx-glr-gallery-lines-bars-and-markers-barchart-py) and

* reverse the order of the bars so the shorter ones are drawn later
* make facecolor and edgecolor the same

we see that the shorter bars now overlap the larger ones, which is not visually appealing.   This does not happen if the edgecolor is left off

image

**Comment by Lucx33** [comment]
@rcomer @timhoffm Removed edgecolor being color by default as you both said. I think this PR is 100% ok now.

**Comment by Lucx33** [comment]
@rcomer @timhoffm Any updates?

**Comment by rcomer** [comment]
@Lucx33 please add a test for the new behaviour.

**Comment by Lucx33** [comment]
@rcomer Is this test good enought?

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2429]:
```suggestion
            The colors of the bar faces. This is an alias for *facecolor*.
            If both are given, *facecolor* takes precedence.
```

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:2432]:
```suggestion
            The colors of the bar faces.
            If both *color* and *facecolor are given, *facecolor* takes precedence.
```

File: lib/matplotlib/tests/test_axes.py
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py:9467]: You don't need to clear in between, which unfortunately is a relatively expensive action. Instead I'd do

```
    # case 1: no color specified
    bars = ax.bar([1, 2, 3], [4, 5, 6])
    [...]

    # case 2: Only 'color'
    bars = ax.bar([11, 12, 13], [4, 5, 6], color='red')
    [...]

    # case 3: Only 'color'
    bars = ax.bar([21, 22, 23], [4, 5, 6], color='red')
    [...]

    # case 4: 'facecolor' and 'color'
    bars = ax.bar([1, 2, 3], [4, 5, 6], color='red', facecolor='green')
    [...]
```

Putting the no color case first ensures noting is messing with the colorcycle.

Optional: You could use https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.same_color.html#matplotlib.colors.same_color for the asserts.

**Comment by Lucx33** [comment]
@timhoffm thx for the review, commited it.

File: lib/matplotlib/tests/test_axes.py
Comment (timhoffm) [inline lib/matplotlib/tests/test_axes.py:9467]:
```suggestion
    bars = ax.bar([11, 12, 13], [4, 5, 6], color='red')
```
and similar for the following plots.

While drawing on top of each other technically works, it's harder to see what's going on when somebody tries to inspect the test result visually, e.g. in case we want to debug an error.

**Review by timhoffm — APPROVED** [review]

**Comment by Lucx33** [comment]
Is there anything else I can do to help with the merge?

**Comment by timhoffm** [comment]
No, we just need a second approval from another core developer.

File: lib/matplotlib/tests/test_axes.py
Comment (rcomer) [inline lib/matplotlib/tests/test_axes.py:9472]: Minor point, but I wonder if we should use a different color than "blue" here.  Since "blue" is the default, this would pass even if the _facecolor_ parameters was ignored.

**Review by rcomer — COMMENTED** [review]
Thanks @Lucx33, I just have one small comment.  Otherwise this looks good to me.

**Comment by Lucx33** [comment]
Messed up my commit but fixed already

**Comment by Lucx33** [comment]
@rcomer Changed blue as you said

**Comment by rcomer** [comment]
Thanks @Lucx33 and congratulations on your first Matplotlib PR!  I hope we hear from you again.

**Pull request merged**
